proposals/p2015.md

```diff
@@ -0,0 +1,270 @@
+# Numeric type literal syntax
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2015)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+    -   [Non-goals](#non-goals)
+-   [Details](#details)
+    -   [Syntax](#syntax)
+    -   [Usage](#usage)
+    -   [Syntax examples and considerations](#syntax-examples-and-considerations)
+        -   [Syntax example use in common fixed-size numeric types](#syntax-example-use-in-common-fixed-size-numeric-types)
+        -   [Syntax example use in arbitrary fixed-size numeric types](#syntax-example-use-in-arbitrary-fixed-size-numeric-types)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [C++ LP64 convention](#c-lp64-convention)
+    -   [Type name with length suffix](#type-name-with-length-suffix)
+    -   [Uppercase suffixes](#uppercase-suffixes)
+    -   [Additional bit sizes](#additional-bit-sizes)
+
+<!-- tocstop -->
+
+## Problem
+
+We want to establish a syntax for fixed-size scalar number types. These types
+include the two's complement signed integer, the unsigned integer, and the
+floating-point number.
+
+As these types are pervasive throughout the language, our goal here is to align
+on a terse, convenient, yet understandable, and ergonomic syntax to the author.
+
+## Background
+
+For developer convenience, names are given to number types that map to native
+machine register widths. These sizes typically include 8-bit, 16-bit, 32-bit,
+64-bit, and, more recently, 128-bit widths.
+
+For example, in [C++11+](https://en.cppreference.com/w/cpp/types/integer),
+integer types such as `int8_t` (8-bit two's complement signed integer type) and
+`uint16_t` (16-bit unsigned integer type) exist, among similar types for 32- and
+64-bit values. Correspondingly, you have the `i8` and `u16`
+([among others](https://doc.rust-lang.org/book/ch03-02-data-types.html#scalar-types))
+scalar integer types in Rust. And in Swift, the `Int8` and `UInt16`
+([among others](https://developer.apple.com/documentation/swift/uint8)) integer
+value types.
+
+In each case, the intent is to provide a clear and pragmatic syntax.
+
+Additional discussion around this proposal's background can be found in
+[#543](https://github.com/carbon-language/carbon-lang/issues/543).
+
+## Proposal
+
+We introduce a simple keyword-like syntax of `iN`, `uN`, and `fN` for two's
```

Comment:
It would be good to say somewhere that we think of these as "type literals" just like `7` is a "numeric literal".